### ISSUE TITLE
Avoid uploading derivations with `allowSubstitutes = false;`

### ISSUE THREAD

**Issue by considerate** [cachix/cachix#513]
Since derivations with `allowSubstitutes = false;` won't be fetched from the binary cache I think that it'd be nice if they never were uploaded to cachix in the first place.

Would it be too much of complexity cost to inspect the derivations before pushing?

**Comment by roberth** [comment]
This needs to access the deriver field of the store output's entry in the store db. (Available in the C++ interface)
The deriver does come with some caveats though. For instance, for fixed output derivations and ca derivations, the deriver may not be the deriver you expect, and for substituted outputs, the deriver may be missing completely. The last one doesn't seem like a problem here though.
Getting the deriver field should be basically free of cost, but reading and parsing the derivation introduces a bit of overhead. That's up to Domen to decide.

**Comment by considerate** [comment]
I was thinking that calling `nix-store --query --binding allowSubstitues <path>` would be sufficient.

```nix
derivation {
  name = "no-subsititute";
  system = "x86_64-linux";
  builder = "builder";
  allowSubstitutes = false;
}
```
```console
$ cat /nix/store/mlay4j5r6396bx9kzb3ki4dayf96998b-no-subsititute.drv
Derive([("out","/nix/store/2d5vcmsjbzjg4hppr4vsvjjvs8mrdpw8-no-subsititute","","")],[],[],"x86_64-linux","builder",[],[("allowSubstitutes",""),("builder","builder"),("name","no-subsititute"),("out","/nix/store/2d5vcmsjbzjg4hppr4vsvjjvs8mrdpw8-no-subsititute"),("system","x86_64-linux")])⏎
$ nix-store --query --binding allowSubstitutes /nix/store/mlay4j5r6396bx9kzb3ki4dayf96998b-no-subsititute.drv

```

The `nix-store` command will exit with a non-zero exitcode if the attribute isn't present in which case I suggest that `cachix` should go ahead and push the derivation to the binary cache.

**Comment by considerate** [comment]
Note that passing a `<path>` that refers to an instantiated package will follow the `deriver` and look up the attribute in the derivation of the package:

```console
$ /nix-store --query --binding allowSubstitutes /nix/store/yq2vv89yivlvzv138biq6gd6mlhgxs44-index-registry-meta-jelly

$ nix-store --query --binding allowSubstitutes /nix/store/ys4dl9p8f47b6lcl9j4ss7x9dfyfw1x2-hello-2.12.1
error: derivation '/nix/store/68ghis80qjda5s9jk8h1cd30n999fp9v-hello-2.12.1.drv' has no environment binding named 'allowSubstitutes'
```

**Comment by domenkozar** [comment]
It's quite complex to support this use case, because you can only filter out those that don't have any ancenstors.

We've recently shipped https://blog.cachix.org/posts/2023-03-08-cachix-1-3-uploads-unleashed/ if you're concerned about upload time, if it's about storage wait a few more weeks :)

**Comment by considerate** [comment]
@domenkozar While I appreciate improvements in upload speed it seems to me that optimistically avoiding to upload files in the first place is a clear improvement over the status quo for the cases you know that you never want to download the files.

I might be misunderstanding your comment but if I understand it correctly you don’t get any false positives with the above proposed check, only false negatives. If that understanding is correct I think the trade off is between having no check and incurring the complexity cost of performing the check for each derivation despite hitting quite rarely.

Personally I lean towards adding the check because it allows me to avoid uploading gigabytes of data unnecessarily to cachix. I do however understand that the use of `allowSubstitues` is rare and most people will only get affected by this through the extra time to check whether the binding is set to `false` or not which incurs a slight penalty on the queue time.

Please correct me if I’m wrong about any of the above. I might patch cachix locally but I think the check might be generally beneficial so I ask you to consider whether the cost is worth it in the average case.

**Comment by considerate** [comment]
On a second reading I see that I might have misunderstood your point.

Is it the case that if you have a derivation `B` that depends on a derivation `A` where `A` has `allowSubstitutes = false` then `A` will be uploaded if you instantiate `B`?

Or is the issue that if `B` has `allowSubstitutes = false` but `A` doesn’t, then `A` will be uploaded to the cache despite you not wanting to download `B` and hence maybe not `A` either? (This was my original reading of your comment)

**Comment by roberth** [comment]
I think the terminology is getting a bit confusing here. I'll add some context:

Cachix generally isn't used on derivations (.drv files) but on outputs of derivations and their closures.

Nix requires that closures are complete. If A depends on B and A is in the store, B must be in the store.

A derivation generally determines the output path (though requires a build for CA derivations), but an output does not determine its deriver: there could be 0 to many.

Reading and parsing a derivation is cheaper than starting a nix subprocess.

-----

Depending on real world performance, perhaps `allowSubstitutes` behavior should be behind a flag? Could be opt-in or opt-out.

**Comment by considerate** [comment]
@roberth Yes, you're quite right. In my above comment please read it as "output" in place of "derivation" everywhere.

> but an output does not determine its deriver: there could be 0 to many.

If there are multiple derivers what would `nix-store --query --binding allowSubstitues <path>` do? With manual testing I only encountered 0 and 1 derivers per output. Would it print the binding for each deriver to stdout?

A more general question, if we cannot determine the deriver of an output could this feature even be implemented? I mean, since `cachix watch-store` is finding [output paths after the fact via the lock files](https://github.com/cachix/cachix/blob/3e53f67a4db937e4dac5a6e2d37744f214f6b1d9/cachix/src/Cachix/Client/WatchStore.hs#L36) how can you tell what derivation to parse in order to figure out if the output path doesn't determine the deriver?

**Comment by roberth** [comment]
The nix database only support 0 or 1 deriver per path and the semantics aren't specified. Nonetheless, there's a good chance that the right thing will happen, because often that one deriver will be good enough and the drv file still present. It's just not possible to provide hard guarantees without specifying certain preconditions. Often those preconditions will be satisfied though, I think.

**Comment by considerate** [comment]
> It's just not possible to provide hard guarantees without specifying certain preconditions. Often those preconditions will be satisfied though, I think.

Yes, I think that the check will miss some potential opportunities to avoid uploading output paths unnecessarily. However as far as I can tell, the opposite is very unlikely to happen, where a path has one deriver with `allowSubsitutes = false` and another without `allowSubstitutes = false` and it's the one with `allowSubstitutes = false` that is referenced by the nix database despite the closure in question was generated with the latter deriver.

Even when that happens, if I understand the behavior of nix correctly the worst thing that will happen is that the path will be recreated by rebuilding (instantiating) the derivation without `allowSubstitutes = false` and the rest will be fetched from cachix.

**Comment by domenkozar** [comment]
I've had a number of interactions with support w.r.t `allowSubstitutes` when someone using Nix gets confused why it's not using the binary cache or vice versa.

My conclusion is that it's [not worth having this complexity](https://github.com/NixOS/nix/issues/4442) and it's a poor man's fix instead of fixing the performance of binary caches.

The complexity is not only in code (which if alone I wouldn't mind) but mainly on the support side: another variable that can go wrong when things aren't working as expected.
